// @marblejs/core/dist/+internal/utils/any.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseJson = exports.stringifyJson = exports.transformUrlEncoded = exports.bufferFrom = exports.wait = exports.getPortEnv = exports.isTestEnv = exports.isNullable = exports.isNonNullable = void 0;
const pipeable_1 = require("fp-ts/lib/pipeable");
const function_1 = require("fp-ts/lib/function");
const O = require("fp-ts/lib/Option");
const E = require("fp-ts/lib/Either");
const qs = require("qs");
const string_util_1 = require("./string.util");
exports.isNonNullable = (value) => value !== undefined && value !== null;
exports.isNullable = (value) => !exports.isNonNullable(value);
exports.isTestEnv = () => process.env.NODE_ENV === 'test';
exports.getPortEnv = () => pipeable_1.pipe(O.fromNullable(process.env.PORT), O.map(Number), O.toUndefined);
exports.wait = (seconds = 1) => new Promise(res => {
    setTimeout(() => res(), seconds * 1000);
});
exports.bufferFrom = (data) => Buffer.from(data);
exports.transformUrlEncoded = (data) => !string_util_1.isString(data) ? qs.stringify(data) : data;
exports.stringifyJson = (data) => !string_util_1.isString(data) ? JSON.stringify(data) : data;
exports.parseJson = (data) => pipeable_1.pipe(E.tryCatch(() => JSON.parse(data), () => data), E.fold(function_1.identity, function_1.identity));
