// @marblejs/core/dist/http/error/http.error.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorNotBoundToRequestErrorFactory = exports.responseNotBoundToRequestErrorFactory = exports.unexpectedErrorWhileSendingOutputFactory = exports.unexpectedErrorWhileSendingErrorFactory = exports.isHttpRequestError = exports.isHttpError = exports.HttpRequestError = exports.HttpError = exports.HttpErrorType = void 0;
const chalk = require("chalk");
const utils_1 = require("../../+internal/utils");
const error_factory_1 = require("../../error/error.factory");
var HttpErrorType;
(function (HttpErrorType) {
    HttpErrorType["HTTP_ERROR"] = "HttpError";
    HttpErrorType["HTTP_REQUEST_ERROR"] = "HttpRequestError";
})(HttpErrorType = exports.HttpErrorType || (exports.HttpErrorType = {}));
class HttpError extends utils_1.NamedError {
    constructor(message, status, data, request, context) {
        super(HttpErrorType.HTTP_ERROR, message);
        this.message = message;
        this.status = status;
        this.data = data;
        this.request = request;
        this.context = context;
    }
}
exports.HttpError = HttpError;
class HttpRequestError extends utils_1.NamedError {
    constructor(request, error) {
        super(HttpErrorType.HTTP_REQUEST_ERROR, 'An error occured while processing a request');
        this.request = request;
        this.error = error;
    }
}
exports.HttpRequestError = HttpRequestError;
exports.isHttpError = (error) => (error === null || error === void 0 ? void 0 : error.name) === HttpErrorType.HTTP_ERROR;
exports.isHttpRequestError = (error) => (error === null || error === void 0 ? void 0 : error.name) === HttpErrorType.HTTP_REQUEST_ERROR;
exports.unexpectedErrorWhileSendingErrorFactory = (error) => {
    const message = `An unexpected error ${chalk.red(`"${error.message}"`)} occured while sending an error response. Please check your error effect.`;
    return error_factory_1.coreErrorFactory(message, { printStacktrace: false });
};
exports.unexpectedErrorWhileSendingOutputFactory = (error) => {
    const message = `An unexpected error ${chalk.red(`"${error.message}"`)} occured while sending a response. Please check your output effect.`;
    return error_factory_1.coreErrorFactory(message, { printStacktrace: false });
};
exports.responseNotBoundToRequestErrorFactory = (response) => {
    const message = `An effect returned a response: "${chalk.yellow(JSON.stringify(response))}" without bound request`;
    return error_factory_1.coreErrorFactory(message, { printStacktrace: false });
};
exports.errorNotBoundToRequestErrorFactory = (error) => {
    const message = `An effect or middleware thrown an error ${chalk.red(`"${error.message}"`)} without bound request.`;
    return error_factory_1.coreErrorFactory(message, { printStacktrace: false });
};
